src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -130,8 +130,12 @@ private unsafe int EncodeIntoBuffer(char* buffer, int bufferLength, char* value,
 
             if (firstCharacterToEncode > 0)
             {
-                int bytesToCopy = firstCharacterToEncode + firstCharacterToEncode;
-                BufferInternal.MemoryCopy(value, buffer, bytesToCopy, bytesToCopy);
+                Debug.Assert(firstCharacterToEncode <= valueLength);
```

Comment:
Presumably, it's guaranteed that `bufferLength >= valueLength >= firstCharacterToEncode`. Maybe add a `Debug.Assert(firstCharacterToEncode <= bufferLength)` (i.e. the `MemoryCopy` is guaranteed to succeed).